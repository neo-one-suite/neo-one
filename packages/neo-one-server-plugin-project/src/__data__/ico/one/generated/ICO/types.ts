// tslint:disable
import {
  AddressString,
  BufferString,
  Event,
  InvokeReceipt,
  InvokeTransactionOptions,
  ReadSmartContract,
  SmartContract,
  TransactionResult,
} from '@neo-one/client';
import BigNumber from 'bignumber.js';

export type ICOEvent = ICOTransferEvent | ICORefundEvent | ICOTraceEvent | ICOErrorEvent | ICOConsoleLogEvent;

export interface ICOTransferEventParameters {
  readonly from: AddressString | undefined;
  readonly to: AddressString | undefined;
  readonly amount: BigNumber;
}
export interface ICOTransferEvent extends Event<'transfer', ICOTransferEventParameters> {}
export interface ICORefundEventParameters {}
export interface ICORefundEvent extends Event<'refund', ICORefundEventParameters> {}
export interface ICOTraceEventParameters {
  readonly line: BigNumber;
}
export interface ICOTraceEvent extends Event<'trace', ICOTraceEventParameters> {}
export interface ICOErrorEventParameters {
  readonly line: BigNumber;
  readonly message: string;
}
export interface ICOErrorEvent extends Event<'error', ICOErrorEventParameters> {}
export interface ICOConsoleLogEventParameters {
  readonly line: BigNumber;
  readonly args: BufferString;
}
export interface ICOConsoleLogEvent extends Event<'console.log', ICOConsoleLogEventParameters> {}

export interface ICOSmartContract extends SmartContract<ICOReadSmartContract> {
  readonly amountPerNEO: () => Promise<BigNumber>;
  readonly deploy: (
    owner?: AddressString,
    startTimeSeconds?: BigNumber,
    icoDurationSeconds?: BigNumber,
    options?: InvokeTransactionOptions,
  ) => Promise<TransactionResult<InvokeReceipt<boolean, ICOEvent>>>;
  readonly icoDurationSeconds: () => Promise<BigNumber>;
  readonly mintTokens: (
    options?: InvokeTransactionOptions,
  ) => Promise<TransactionResult<InvokeReceipt<boolean, ICOEvent>>>;
  readonly owner: () => Promise<AddressString>;
  readonly remaining: () => Promise<BigNumber>;
  readonly startTimeSeconds: () => Promise<BigNumber>;
}

export interface ICOReadSmartContract extends ReadSmartContract<ICOEvent> {
  readonly amountPerNEO: () => Promise<BigNumber>;
  readonly icoDurationSeconds: () => Promise<BigNumber>;
  readonly owner: () => Promise<AddressString>;
  readonly remaining: () => Promise<BigNumber>;
  readonly startTimeSeconds: () => Promise<BigNumber>;
}
