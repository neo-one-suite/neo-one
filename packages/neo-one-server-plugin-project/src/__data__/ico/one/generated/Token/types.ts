// tslint:disable
import {
  AddressString,
  BufferString,
  Event,
  InvokeReceipt,
  InvokeTransactionOptions,
  ReadSmartContract,
  SmartContract,
  TransactionResult,
} from '@neo-one/client';
import BigNumber from 'bignumber.js';

export type TokenEvent = TokenTransferEvent | TokenTraceEvent | TokenErrorEvent | TokenConsoleLogEvent;

export interface TokenTransferEventParameters {
  readonly from: AddressString | undefined;
  readonly to: AddressString | undefined;
  readonly amount: BigNumber;
}
export interface TokenTransferEvent extends Event<'transfer', TokenTransferEventParameters> {}
export interface TokenTraceEventParameters {
  readonly line: BigNumber;
}
export interface TokenTraceEvent extends Event<'trace', TokenTraceEventParameters> {}
export interface TokenErrorEventParameters {
  readonly line: BigNumber;
  readonly message: string;
}
export interface TokenErrorEvent extends Event<'error', TokenErrorEventParameters> {}
export interface TokenConsoleLogEventParameters {
  readonly line: BigNumber;
  readonly args: BufferString;
}
export interface TokenConsoleLogEvent extends Event<'console.log', TokenConsoleLogEventParameters> {}

export interface TokenSmartContract extends SmartContract<TokenReadSmartContract> {
  readonly balanceOf: (address: AddressString) => Promise<BigNumber>;
  readonly decimals: () => Promise<BigNumber>;
  readonly deploy: (
    owner?: AddressString,
    options?: InvokeTransactionOptions,
  ) => Promise<TransactionResult<InvokeReceipt<boolean, TokenEvent>>>;
  readonly issue: (
    to: AddressString,
    amount: BigNumber,
    options?: InvokeTransactionOptions,
  ) => Promise<TransactionResult<InvokeReceipt<boolean, TokenEvent>>>;
  readonly name: () => Promise<string>;
  readonly owner: () => Promise<AddressString>;
  readonly symbol: () => Promise<string>;
  readonly totalSupply: () => Promise<BigNumber>;
  readonly transfer: (
    from: AddressString,
    to: AddressString,
    amount: BigNumber,
    options?: InvokeTransactionOptions,
  ) => Promise<TransactionResult<InvokeReceipt<boolean, TokenEvent>>>;
}

export interface TokenReadSmartContract extends ReadSmartContract<TokenEvent> {
  readonly balanceOf: (address: AddressString) => Promise<BigNumber>;
  readonly decimals: () => Promise<BigNumber>;
  readonly name: () => Promise<string>;
  readonly owner: () => Promise<AddressString>;
  readonly symbol: () => Promise<string>;
  readonly totalSupply: () => Promise<BigNumber>;
}
