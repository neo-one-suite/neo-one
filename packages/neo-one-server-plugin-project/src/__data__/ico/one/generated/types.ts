// tslint:disable
import { ICOSmartContract } from './ICO/types';
import { TokenSmartContract } from './Token/types';

export interface Contracts {
  readonly ico: ICOSmartContract;
  readonly token: TokenSmartContract;
}
