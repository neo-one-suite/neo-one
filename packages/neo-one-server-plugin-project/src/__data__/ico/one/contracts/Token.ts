import {
  Address,
  constant,
  Contract,
  createEventNotifier,
  Fixed,
  MapStorage,
  SmartContract,
} from '@neo-one/smart-contract';

const notifyTransfer = createEventNotifier<Address | undefined, Address | undefined, Fixed<8>>(
  'transfer',
  'from',
  'to',
  'amount',
);

export class Token implements SmartContract {
  public readonly properties = {
    codeVersion: '1.0',
    author: 'dicarlo2',
    email: 'alex.dicarlo@neotracker.io',
    description: 'NEOâ€¢ONE Token',
    payable: false,
  };
  public readonly name = 'One';
  public readonly symbol = 'ONE';
  public readonly decimals = 8;
  private mutableSupply: Fixed<8> = 0;
  private readonly balances = new MapStorage<Address, Fixed<8>>();

  public constructor(public readonly owner: Address = Address.from('AXNajBTQLxWHwc9sKyXcc4UdbJvp3arYDG')) {
    if (!Address.verifySender(owner)) {
      throw new Error('Sender was not the owner.');
    }
  }

  @constant
  public get totalSupply(): Fixed<8> {
    return this.mutableSupply;
  }

  @constant
  public balanceOf(address: Address): Fixed<8> {
    const balance = this.balances.get(address);

    return balance === undefined ? 0 : balance;
  }

  public transfer(from: Address, to: Address, amount: Fixed<8>): boolean {
    if (amount < 0) {
      throw new Error(`Amount must be greater than 0: ${amount}`);
    }

    if (!Address.verifySender(from)) {
      return false;
    }

    const contract = Contract.for(to);
    if (contract !== undefined && !contract.payable) {
      return false;
    }

    const fromBalance = this.balanceOf(from);
    if (fromBalance < amount) {
      return false;
    }

    const toBalance = this.balanceOf(to);
    this.balances.set(from, fromBalance - amount);
    this.balances.set(to, toBalance + amount);
    notifyTransfer(from, to, amount);

    return true;
  }

  public issue(to: Address, amount: Fixed<8>): boolean {
    if (amount < 0) {
      throw new Error(`Amount must be greater than 0: ${amount}`);
    }

    const toBalance = this.balanceOf(to);
    this.balances.set(to, toBalance + amount);
    notifyTransfer(undefined, to, amount);
    this.mutableSupply += amount;

    return true;
  }
}
