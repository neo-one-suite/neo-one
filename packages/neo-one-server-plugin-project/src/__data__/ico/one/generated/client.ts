/* @hash 83bfde457ae311a31347a44614332db3 */
// tslint:disable
/* eslint-disable */
import {
  Client,
  DeveloperClient,
  LocalClient,
  LocalKeyStore,
  LocalMemoryStore,
  LocalUserAccountProvider,
  NEOONEProvider,
  NEOONEOneDataProvider,
  OneClient,
  NEOONEDataProviderOptions,
  UserAccountProvider,
  UserAccountProviders,
} from '@neo-one/client';
import { projectID } from './projectID';

export type DefaultUserAccountProviders = {
  readonly memory: LocalUserAccountProvider<LocalKeyStore, NEOONEProvider>;
};
const getDefaultUserAccountProviders = (provider: NEOONEProvider): DefaultUserAccountProviders => ({
  memory: new LocalUserAccountProvider({
    keystore: new LocalKeyStore({ store: new LocalMemoryStore() }),
    provider,
  }),
});

const isLocalUserAccountProvider = (
  userAccountProvider: UserAccountProvider,
): userAccountProvider is LocalUserAccountProvider<any, any> => userAccountProvider instanceof LocalUserAccountProvider;

export const createClient = <TUserAccountProviders extends UserAccountProviders<any> = DefaultUserAccountProviders>(
  getUserAccountProviders: (provider: NEOONEProvider) => TUserAccountProviders = getDefaultUserAccountProviders as any,
): Client<
  TUserAccountProviders extends UserAccountProviders<infer TUserAccountProvider> ? TUserAccountProvider : never,
  TUserAccountProviders
> => {
  const providers: Array<NEOONEOneDataProvider | NEOONEDataProviderOptions> = [];
  if (process.env.NODE_ENV !== 'production' || process.env.NEO_ONE_DEV === 'true') {
    providers.push(new NEOONEOneDataProvider({ network: 'local', projectID, port: 48242 }));
  }
  const provider = new NEOONEProvider(providers);

  const userAccountProviders = getUserAccountProviders(provider);
  const localUserAccountProviders = Object.values(userAccountProviders).filter(isLocalUserAccountProvider);
  const localUserAccountProvider = localUserAccountProviders.find(
    (userAccountProvider) => userAccountProvider.keystore instanceof LocalKeyStore,
  );

  if (process.env.NODE_ENV !== 'production' || process.env.NEO_ONE_DEV === 'true') {
    if (localUserAccountProvider !== undefined) {
      const localKeyStore = localUserAccountProvider.keystore;
      if (localKeyStore instanceof LocalKeyStore) {
        Promise.all([
          localKeyStore.addAccount({
            network: 'local',
            name: 'master',
            privateKey: 'L4qhHtwbiAMu1nrSmsTP5a3dJbxA3SNS6oheKnKd8E7KTJyCLcUv',
          }),
          localKeyStore.addAccount({
            network: 'local',
            name: 'alfa',
            privateKey: 'KyX5sPKRpAMb3XAFLUrHv7u1LxKkKFrpyJDgE4kceRX9FRJ4WRCQ',
          }),
          localKeyStore.addAccount({
            network: 'local',
            name: 'bravo',
            privateKey: 'L5LfJc2Ngsxu8ZFMvnJbYJ1QdQCstzRXmKLybsFS1aQFURkJ5CHS',
          }),
          localKeyStore.addAccount({
            network: 'local',
            name: 'charlie',
            privateKey: 'KxCH2Ei4TLqp2Qa7swz9bstQc5uREiCpvzvL9R6xLX8X5U8ZqeBj',
          }),
          localKeyStore.addAccount({
            network: 'local',
            name: 'delta',
            privateKey: 'KyVvngWhhfHiociMuwyLmGw8xTu9myKXRnvv5Fes9jDMa2Zyc6P9',
          }),
          localKeyStore.addAccount({
            network: 'local',
            name: 'echo',
            privateKey: 'L37qr7PWqWmgjUPfRC9mS78GjRxgGi4azySCsLUBMAa5hMka2JEm',
          }),
          localKeyStore.addAccount({
            network: 'local',
            name: 'foxtrot',
            privateKey: 'KwFf8gdSWxvC5Pp8AidNdF6mHqjH3CukyF3RnfwS5vzMQKLGTP13',
          }),
          localKeyStore.addAccount({
            network: 'local',
            name: 'golf',
            privateKey: 'Kyn2BN3QuHGYgkt9qJgvwzY8yH4xgTUAKwnGhvU1w8Nh3JnivrAr',
          }),
          localKeyStore.addAccount({
            network: 'local',
            name: 'hotel',
            privateKey: 'L5UXfz1xyzDkghGwistNMCV8pbpU4fg14Ez9rfo1y4KgwiadnWX3',
          }),
          localKeyStore.addAccount({
            network: 'local',
            name: 'india',
            privateKey: 'L5Yoq3X4ojx2FvZZxHbMcvT6var4LaXKHEpMYyyxw4jjhSUNJTRa',
          }),
          localKeyStore.addAccount({
            network: 'local',
            name: 'juliett',
            privateKey: 'L1DWex8PtmQJH4GYK9YAuyzmotyL6anY937LxJF54iaALrTtxsD6',
          }),
        ]).catch(() => {
          // do nothing
        });
      }
    }
  }
  return new Client(userAccountProviders);
};

export const createDeveloperClients = (): { [network: string]: DeveloperClient } => ({
  local: new DeveloperClient(new NEOONEOneDataProvider({ network: 'local', projectID, port: 48242 })),
});

export const createLocalClients = (): { [network: string]: LocalClient } => {
  const client = new OneClient(48242);
  return {
    local: {
      getNEOTrackerURL: async () => {
        const result = await client.request({
          plugin: '@neo-one/server-plugin-project',
          options: { type: 'neotracker', projectID },
        });

        return result.response;
      },
      reset: async () => {
        await client.executeTaskList({
          plugin: '@neo-one/server-plugin-project',
          options: {
            command: 'reset',
            projectID,
          },
        });
      },
    },
  };
};
