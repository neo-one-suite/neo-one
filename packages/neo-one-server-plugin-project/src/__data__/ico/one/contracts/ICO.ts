import {
  Address,
  Blockchain,
  constant,
  createEventNotifier,
  Fixed,
  Hash256,
  Integer,
  LinkedSmartContract,
  SmartContract,
  verify,
} from '@neo-one/smart-contract';
import { Token } from './Token';

const notifyRefund = createEventNotifier('refund');

export class ICO implements SmartContract {
  public readonly properties = {
    codeVersion: '1.0',
    author: 'dicarlo2',
    email: 'alex.dicarlo@neotracker.io',
    description: 'NEOâ€¢ONE ICO',
    payable: true,
  };
  public readonly amountPerNEO = 10;
  private mutableRemaining: Fixed<8> = 10_000_000_000_00000000;

  public constructor(
    public readonly owner: Address = Address.from('AXNajBTQLxWHwc9sKyXcc4UdbJvp3arYDG'),
    public readonly startTimeSeconds: Integer = 1534108415,
    public readonly icoDurationSeconds: Integer = 157700000,
  ) {
    if (!Address.verifySender(owner)) {
      throw new Error('Sender was not the owner.');
    }
  }

  @constant
  public get remaining(): Fixed<8> {
    return this.mutableRemaining;
  }

  @verify
  public mintTokens(): boolean {
    if (!this.hasStarted() || this.hasEnded()) {
      notifyRefund();

      return false;
    }

    const { references } = Blockchain.currentTransaction;
    if (references.length === 0) {
      return false;
    }
    const sender = references[0].address;

    let amount = 0;
    // tslint:disable-next-line no-loop-statement
    for (const output of Blockchain.currentTransaction.outputs) {
      if (output.address.equals(Blockchain.contractAddress)) {
        if (!output.asset.equals(Hash256.NEO)) {
          notifyRefund();

          return false;
        }

        amount += output.value * this.amountPerNEO;
      }
    }

    if (amount > this.remaining) {
      notifyRefund();

      return false;
    }

    if (amount === 0) {
      return false;
    }

    const token = LinkedSmartContract.for<Token>();
    if (token.issue(sender, amount)) {
      this.mutableRemaining -= amount;

      return true;
    }

    notifyRefund();

    return false;
  }

  private hasStarted(): boolean {
    return Blockchain.currentBlockTime >= this.startTimeSeconds;
  }

  private hasEnded(): boolean {
    return Blockchain.currentBlockTime > this.startTimeSeconds + this.icoDurationSeconds;
  }
}
