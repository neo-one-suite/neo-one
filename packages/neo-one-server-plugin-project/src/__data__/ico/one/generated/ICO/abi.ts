// tslint:disable
import { ABI } from '@neo-one/client';

export const icoABI: ABI = {
  events: [
    {
      name: 'transfer',
      parameters: [
        {
          name: 'from',
          optional: true,
          type: 'Address',
        },
        {
          name: 'to',
          optional: true,
          type: 'Address',
        },
        {
          decimals: 8,
          name: 'amount',
          optional: false,
          type: 'Integer',
        },
      ],
    },
    {
      name: 'refund',
      parameters: [],
    },
    {
      name: 'trace',
      parameters: [
        {
          decimals: 0,
          name: 'line',
          type: 'Integer',
        },
      ],
    },
    {
      name: 'error',
      parameters: [
        {
          decimals: 0,
          name: 'line',
          type: 'Integer',
        },
        {
          name: 'message',
          type: 'String',
        },
      ],
    },
    {
      name: 'console.log',
      parameters: [
        {
          decimals: 0,
          name: 'line',
          type: 'Integer',
        },
        {
          name: 'args',
          type: 'Buffer',
        },
      ],
    },
  ],
  functions: [
    {
      constant: true,
      name: 'amountPerNEO',
      returnType: {
        decimals: 0,
        optional: false,
        type: 'Integer',
      },
      verify: false,
    },
    {
      constant: true,
      name: 'owner',
      returnType: {
        optional: false,
        type: 'Address',
      },
      verify: false,
    },
    {
      constant: true,
      name: 'startTimeSeconds',
      returnType: {
        decimals: 0,
        optional: false,
        type: 'Integer',
      },
      verify: false,
    },
    {
      constant: true,
      name: 'icoDurationSeconds',
      returnType: {
        decimals: 0,
        optional: false,
        type: 'Integer',
      },
      verify: false,
    },
    {
      constant: true,
      name: 'remaining',
      returnType: {
        decimals: 8,
        optional: false,
        type: 'Integer',
      },
      verify: false,
    },
    {
      constant: false,
      name: 'mintTokens',
      parameters: [],
      returnType: {
        optional: false,
        type: 'Boolean',
      },
      verify: true,
    },
    {
      constant: false,
      name: 'deploy',
      parameters: [
        {
          name: 'owner',
          optional: true,
          type: 'Address',
        },
        {
          decimals: 0,
          name: 'startTimeSeconds',
          optional: true,
          type: 'Integer',
        },
        {
          decimals: 0,
          name: 'icoDurationSeconds',
          optional: true,
          type: 'Integer',
        },
      ],
      returnType: {
        type: 'Boolean',
      },
      verify: false,
    },
  ],
};
