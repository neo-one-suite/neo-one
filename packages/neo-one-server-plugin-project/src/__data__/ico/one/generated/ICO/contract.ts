// tslint:disable
import { Client, ReadClient, SmartContractDefinition } from '@neo-one/client';
import { icoABI } from './abi';
import { ICOReadSmartContract, ICOSmartContract } from './types';
import { sourceMaps } from '../sourceMaps';

const definition: SmartContractDefinition = {
  networks: {
    local: {
      address: 'AQwrDxLW4ciZW82yi8HnPBHthwh6JZJoJn',
    },
  },
  abi: icoABI,
  sourceMaps,
};

export const createICOSmartContract = (client: Client): ICOSmartContract =>
  client.smartContract<ICOSmartContract>(definition);

export const createICOReadSmartContract = (client: ReadClient): ICOReadSmartContract =>
  client.smartContract<ICOReadSmartContract>({
    address: definition.networks[client.dataProvider.network].address,
    abi: definition.abi,
    sourceMaps: definition.sourceMaps,
  });
