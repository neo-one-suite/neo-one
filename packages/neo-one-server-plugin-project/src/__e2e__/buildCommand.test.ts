import {
  ABI,
  Client,
  Contract,
  DeveloperClient,
  Hash256,
  LocalKeyStore,
  LocalMemoryStore,
  LocalUserAccountProvider,
  NEOONEProvider,
  privateKeyToAddress,
  ReadClient,
  UserAccountID,
  wifToPrivateKey,
} from '@neo-one/client';
import { common, crypto } from '@neo-one/client-core';
import { Network } from '@neo-one/server-plugin-network';
import { Wallet } from '@neo-one/server-plugin-wallet';
import { TestOptions } from '@neo-one/smart-contract-test';
import BigNumber from 'bignumber.js';
import * as path from 'path';
import { ICOReadSmartContract, ICOSmartContract } from '../__data__/ico/one/generated/ICO/types';
import { TokenReadSmartContract, TokenSmartContract } from '../__data__/ico/one/generated/Token/types';
import { Contracts } from '../__data__/ico/one/generated/types';

const TO_PRIVATE_KEY = '7d128a6d096f0c14c3a25a2b0c41cf79661bfcb4a8cc95aaaea28bde4d732344';
const TO_PUBLIC_KEY = '02028a99826edc0c97d18e22b6932373d908d323aa7f92656a77ec26e8861699ef';

const LOCAL = 'local';

const getNetworks = async (): Promise<ReadonlyArray<Network>> => {
  const networksString = await one.execute('get network --json');

  return JSON.parse(networksString);
};

const getWallets = async (networkName: string): Promise<ReadonlyArray<Wallet>> => {
  const walletsString = await one.execute(`get wallet --json --network ${networkName}`);

  return JSON.parse(walletsString);
};

const getClients = async (
  network: Network,
  wallet: Wallet,
): Promise<{
  readonly client: Client;
  readonly developerClient: DeveloperClient;
}> => {
  const keystore = new LocalKeyStore({ store: new LocalMemoryStore() });

  if (wallet.wif === undefined) {
    throw new Error(`Something went wrong, wif is null for ${wallet.name}`);
  }
  await keystore.addAccount({
    network: LOCAL,
    name: wallet.name,
    privateKey: wifToPrivateKey(wallet.wif),
  });

  const provider = new NEOONEProvider([{ network: LOCAL, rpcURL: network.nodes[0].rpcAddress }]);
  const localUserAccountProvider = new LocalUserAccountProvider({ keystore, provider });
  const client = new Client({ memory: localUserAccountProvider });

  const developerClient = new DeveloperClient(provider.read(LOCAL));

  return { client, developerClient };
};

const getContracts = async (client: Client, networkName: string): Promise<ReadonlyArray<Contract>> => {
  const readClient = client.read(networkName);
  const blockHeight = await readClient.getBlockCount();

  const mutableContracts: Contract[] = [];
  // tslint:disable-next-line no-loop-statement
  for await (const block of readClient.iterBlocks({ indexStop: blockHeight })) {
    // tslint:disable-next-line no-loop-statement
    for (const transaction of block.transactions) {
      if (transaction.type === 'InvocationTransaction' && transaction.invocationData.contracts.length > 0) {
        mutableContracts.push(transaction.invocationData.contracts[0]);
      }
    }
  }

  return mutableContracts;
};

const getGeneratedICOCode = (): {
  readonly abi: ABI;
  readonly contract: {
    readonly createSmartContract: (client: Client) => ICOSmartContract;
    readonly createReadSmartContract: (readClient: ReadClient) => ICOReadSmartContract;
  };
} => {
  // tslint:disable-next-line no-require-imports
  const abi = require('../__data__/ico/one/generated/ICO/abi').icoABI;
  // tslint:disable-next-line no-require-imports
  const contract = require('../__data__/ico/one/generated/ICO/contract');
  const createSmartContract = contract.createICOSmartContract;
  const createReadSmartContract = contract.createICOReadSmartContract;

  return { abi, contract: { createSmartContract, createReadSmartContract } };
};

const getGeneratedTokenCode = (): {
  readonly abi: ABI;
  readonly contract: {
    readonly createSmartContract: (client: Client) => TokenSmartContract;
    readonly createReadSmartContract: (readClient: ReadClient) => TokenReadSmartContract;
  };
} => {
  // tslint:disable-next-line no-require-imports
  const abi = require('../__data__/ico/one/generated/Token/abi').tokenABI;
  // tslint:disable-next-line no-require-imports
  const contract = require('../__data__/ico/one/generated/Token/contract');
  const createSmartContract = contract.createTokenSmartContract;
  const createReadSmartContract = contract.createTokenReadSmartContract;

  return { abi, contract: { createSmartContract, createReadSmartContract } };
};

const verifySmartContractAfterMint = async (
  developerClient: DeveloperClient,
  ico: ICOSmartContract,
  token: TokenSmartContract,
  accountID: UserAccountID,
  toAccountID: UserAccountID,
): Promise<void> => {
  const [totalSupply, remaining, balance, toBalance] = await Promise.all([
    token.totalSupply(),
    ico.remaining(),
    token.balanceOf(accountID.address),
    token.balanceOf(toAccountID.address),
  ]);
  expect(totalSupply.toString()).toEqual('100');
  expect(remaining.toString()).toEqual(new BigNumber(9_999_999_900).toString());
  expect(balance.toString()).toEqual('100');
  expect(toBalance.toString()).toEqual('0');

  const result = await token.transfer(accountID.address, toAccountID.address, new BigNumber('25'));
  const [receipt] = await Promise.all([result.confirmed({ timeoutMS: 2500 }), developerClient.runConsensusNow()]);
  if (receipt.result.state === 'FAULT') {
    throw new Error(receipt.result.message);
  }

  const [totalSupplyAfter, balanceAfter, toBalanceAfter] = await Promise.all([
    token.totalSupply(),
    token.balanceOf(accountID.address),
    token.balanceOf(toAccountID.address),
  ]);
  expect(totalSupplyAfter.toString()).toEqual('100');
  expect(balanceAfter.toString()).toEqual('75');
  expect(toBalanceAfter.toString()).toEqual('25');
};

const verifySmartContracts = async (
  developerClient: DeveloperClient,
  ico: ICOSmartContract,
  token: TokenSmartContract,
  icoRead: ICOReadSmartContract,
  tokenRead: TokenReadSmartContract,
  accountID: UserAccountID,
  toAccountID: UserAccountID,
): Promise<void> => {
  const [
    name,
    symbol,
    decimals,
    amountPerNEO,
    icoOwner,
    owner,
    startTimeSeconds,
    icoDurationSeconds,
    initialTotalSupply,
    [initialRemaining, initialBalance],
  ] = await Promise.all([
    tokenRead.name(),
    tokenRead.symbol(),
    tokenRead.decimals(),
    icoRead.amountPerNEO(),
    icoRead.owner(),
    tokenRead.owner(),
    icoRead.startTimeSeconds(),
    icoRead.icoDurationSeconds(),
    tokenRead.totalSupply(),
    Promise.all([icoRead.remaining(), tokenRead.balanceOf(accountID.address)]),
  ]);
  expect(name).toEqual('One');
  expect(symbol).toEqual('ONE');
  expect(decimals.toString()).toEqual('8');
  expect(amountPerNEO.toString()).toEqual('10');
  expect(icoOwner).toEqual(accountID.address);
  expect(owner).toEqual(accountID.address);
  expect(startTimeSeconds.toString()).toEqual('1534108415');
  expect(icoDurationSeconds.toString()).toEqual('157700000');
  expect(initialTotalSupply.toString()).toEqual('0');
  expect(initialRemaining.toString()).toEqual(new BigNumber(10_000_000_000).toString());
  expect(initialBalance.toString()).toEqual('0');

  const mintResult = await ico.mintTokens({
    transfers: [
      {
        to: ico.definition.networks[accountID.network].address,
        amount: new BigNumber(10),
        asset: Hash256.NEO,
      },
    ],
  });

  const [mintReceipt] = await Promise.all([mintResult.confirmed(), developerClient.runConsensusNow()]);
  if (mintReceipt.result.state === 'FAULT') {
    throw new Error(mintReceipt.result.message);
  }

  expect(mintReceipt.result.state).toEqual('HALT');
  expect(mintReceipt.result.value).toEqual(true);
  expect(mintReceipt.result.gasCost).toMatchSnapshot('mint cost');
  expect(mintReceipt.result.gasConsumed).toMatchSnapshot('mint consumed');
  expect(mintReceipt.events).toHaveLength(1);
  const event = mintReceipt.events[0];
  expect(event.name).toEqual('transfer');
  if (event.name !== 'transfer') {
    throw new Error('For TS');
  }
  expect(event.parameters.from).toBeUndefined();
  expect(event.parameters.to).toEqual(accountID.address);
  expect(event.parameters.amount.toString()).toEqual('100');

  await verifySmartContractAfterMint(developerClient, ico, token, accountID, toAccountID);
};

type WithContracts = (test: (contracts: Contracts & TestOptions) => Promise<void>) => Promise<void>;
const verifySmartContractsTest = async () => {
  // tslint:disable-next-line no-require-imports
  const test = require('../__data__/ico/one/generated/test');
  const withContracts: WithContracts = test.withContracts;
  await withContracts(async ({ ico, token, developerClient, masterAccountID, networkName }) => {
    await verifySmartContracts(
      developerClient,
      ico,
      token,
      ico.read(networkName),
      token.read(networkName),
      masterAccountID,
      { network: networkName, address: privateKeyToAddress(TO_PRIVATE_KEY) },
    );
  });
};

const verifySmartContractsManual = async (
  client: Client,
  developerClient: DeveloperClient,
  accountID: UserAccountID,
  toAccountID: UserAccountID,
) => {
  const {
    abi: icoABI,
    contract: { createSmartContract: createICOSmartContract, createReadSmartContract: createICOReadSmartContract },
  } = getGeneratedICOCode();
  expect(icoABI).toBeDefined();
  const {
    abi: tokenABI,
    contract: { createSmartContract: createTokenSmartContract, createReadSmartContract: createTokenReadSmartContract },
  } = getGeneratedTokenCode();
  expect(tokenABI).toBeDefined();

  const ico = createICOSmartContract(client);
  const icoRead = createICOReadSmartContract(client.read(accountID.network));
  const token = createTokenSmartContract(client);
  const tokenRead = createTokenReadSmartContract(client.read(accountID.network));

  await verifySmartContracts(developerClient, ico, token, icoRead, tokenRead, accountID, toAccountID);
};

const verifyICOContract = (contract?: Contract): void => {
  expect(contract).toBeDefined();
  if (contract === undefined) {
    throw new Error('For TS');
  }
  expect(contract.codeVersion).toEqual('1.0');
  expect(contract.author).toEqual('dicarlo2');
  expect(contract.email).toEqual('alex.dicarlo@neotracker.io');
  expect(contract.description).toEqual('NEO•ONE ICO');
  expect(contract.parameters).toEqual(['String', 'Array']);
  expect(contract.returnType).toEqual('Buffer');
  expect(contract.payable).toBeTruthy();
  expect(contract.storage).toBeTruthy();
  expect(contract.dynamicInvoke).toBeFalsy();
};

const verifyTokenContract = (contract?: Contract): void => {
  expect(contract).toBeDefined();
  if (contract === undefined) {
    throw new Error('For TS');
  }
  expect(contract.codeVersion).toEqual('1.0');
  expect(contract.author).toEqual('dicarlo2');
  expect(contract.email).toEqual('alex.dicarlo@neotracker.io');
  expect(contract.description).toEqual('NEO•ONE Token');
  expect(contract.parameters).toEqual(['String', 'Array']);
  expect(contract.returnType).toEqual('Buffer');
  expect(contract.payable).toBeFalsy();
  expect(contract.storage).toBeTruthy();
  expect(contract.dynamicInvoke).toBeFalsy();
};

describe('buildCommand', () => {
  test('build', async () => {
    crypto.addPublicKey(common.stringToPrivateKey(TO_PRIVATE_KEY), common.stringToECPoint(TO_PUBLIC_KEY));

    await one.execute('build', { cwd: path.resolve(__dirname, '..', '__data__', 'ico') });

    const networks = await getNetworks();
    expect(networks.length).toEqual(1);

    const network = networks[0];
    expect(network.name).toEqual('ico-local');
    expect(network.type).toEqual('private');
    expect(network.nodes.length).toEqual(1);
    expect(network.nodes[0].live).toBeTruthy();
    expect(network.nodes[0].ready).toBeTruthy();

    const wallets = await getWallets(LOCAL);
    expect(wallets.length).toEqual(1);
    const wallet = wallets[0];

    const { client, developerClient } = await getClients(network, wallet);

    const contracts = await getContracts(client, LOCAL);
    verifyICOContract(contracts.find((contract) => contract.name === 'ICO'));
    verifyTokenContract(contracts.find((contract) => contract.name === 'Token'));

    await Promise.all([
      verifySmartContractsTest(),
      verifySmartContractsManual(
        client,
        developerClient,
        { ...wallet.accountID, network: LOCAL },
        { network: LOCAL, address: privateKeyToAddress(TO_PRIVATE_KEY) },
      ),
    ]);

    await one.execute('build', { cwd: path.resolve(__dirname, '..', '__data__', 'ico') });
  });
});
