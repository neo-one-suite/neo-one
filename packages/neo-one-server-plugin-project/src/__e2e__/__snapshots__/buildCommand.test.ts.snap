// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildCommand build: mint consumed 1`] = `"0"`;

exports[`buildCommand build: mint consumed 2`] = `"0"`;

exports[`buildCommand build: mint cost 1`] = `"9.92"`;

exports[`buildCommand build: mint cost 2`] = `"9.92"`;
